File: js/24Game.js
/*
You have 4 cards each containing a number from 1 to 9. You need to judge whether they could operated through *, /, +, -, (, ) to get the value of 24.

Example 1:
Input: [4, 1, 8, 7]
Output: True
Explanation: (8-4) * (7-1) = 24
Example 2:
Input: [1, 2, 1, 2]
Output: False
Note:
The division operator / represents real division, not integer division. For example, 4 / (1 - 2/3) = 12.
Every operation done is between two numbers. In particular, we cannot use - as a unary operator. For example, with [1, 1, 1, 1] as input, the expression -1 - 1 - 1 - 1 is not allowed.
You cannot concatenate numbers together. For example, if the input is [1, 2, 1, 2], we cannot write this as 12 + 12. */
/**
 * @param {number[]} nums
 * @return {boolean}
 */

/**
 * Find all the possible combinations (since there are only 4 numbers so the combinations are limited)
 */
var judgePoint24 = function(nums) {
  if (nums.length === 1) {
    return Math.abs(24 - nums[0]) < 0.0001; // Approx towards 24
  }

  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      let combinations = calculate(nums[i], nums[j]);
      let arr = nums.filter((v, idx) => idx !== j && idx !== i);
      for (let c of combinations) {
        arr.push(c);
        if (judgePoint24(arr)) {
          return true;
        }
        arr.pop();
      }
    }
  }
  return false;
};

var calculate = function(a, b) {
  // Calculate all the possible combinations
  let res = [a * b, a + b, a - b, b - a];
  if (b > 0) {
    res.push(a / b);
  }
  if (a > 0) {
    res.push(b / a);
  }
  return res;
};

File: py/24game.py
class Solution:
    def judgePoint24(self, nums: List[int]) -> bool:
        if len(nums) == 1:
            return abs(nums[0] - 24) < 0.0001
        n = len(nums)
        for i in range(n):
            for j in range(i+1, n):
                nextNums = []
                for k in range(n):
                    if k != i and k != j:
                        nextNums.append(nums[k])
                combinations = self.compute(nums[i], nums[j])
                for c in combinations:
                    if(self.judgePoint24(nextNums+[c])):
                        return True
        return False

    def compute(self, num1, num2):
        res = [num1+num2, num1*num2, num1-num2, num2-num1]
        if num1 != 0:
            res.append(num2/num1)
        if num2 != 0:
            res.append(num1/num2)
        return res